src/libpocketsphinx/dict2pid.c

```diff
@@ -401,96 +403,135 @@ dict2pid_build(bin_mdef_t * mdef, dict_t * dict)
-                                     mdef->n_ciphone, sizeof(s3ssid_t));
-
-    dict2pid->lrdiph_rc = (s3ssid_t ***) ckd_calloc_3d(mdef->n_ciphone,
-                                                       mdef->n_ciphone,
-                                                       mdef->n_ciphone,
-                                                       sizeof
-                                                       (s3ssid_t));
-    /* Actually could use memset for this, if BAD_S3SSID is guaranteed
-     * to be 65535... */
-    for (b = 0; b < mdef->n_ciphone; ++b) {
-        for (r = 0; r < mdef->n_ciphone; ++r) {
-            for (l = 0; l < mdef->n_ciphone; ++l) {
-                dict2pid->ldiph_lc[b][r][l] = BAD_S3SSID;
-                dict2pid->lrdiph_rc[b][l][r] = BAD_S3SSID;
-                rdiph_rc[b][l][r] = BAD_S3SSID;
-            }
-        }
-    }
 
-    /* Track which diphones / ciphones have been seen. */
-    ldiph = bitvec_alloc(mdef->n_ciphone * mdef->n_ciphone);
-    rdiph = bitvec_alloc(mdef->n_ciphone * mdef->n_ciphone);
-    single = bitvec_alloc(mdef->n_ciphone);
+    dict2pid->ci_mode = (mdef->n_ciphone == mdef->n_phone);
 
-    for (w = 0; w < dict_size(dict2pid->dict); w++) {
-        pronlen = dict_pronlen(dict, w);
+    if (dict2pid->ci_mode) {
+        dict2pid->ldiph_lc = (s3ssid_t ***) ckd_calloc_3d(mdef->n_ciphone, 1, 1, sizeof(s3ssid_t));
+        dict2pid->lrdiph_rc = (s3ssid_t ***) ckd_calloc_3d(mdef->n_ciphone, 1, 1, sizeof(s3ssid_t));
 
-        if (pronlen >= 2) {
-            b = dict_first_phone(dict, w);
-            r = dict_second_phone(dict, w);
-            /* Populate ldiph_lc */
-            if (bitvec_is_clear(ldiph, b * mdef->n_ciphone + r)) {
-                /* Mark this diphone as done */
-                bitvec_set(ldiph, b * mdef->n_ciphone + r);
-
-                /* Record all possible ssids for b(?,r) */
-                for (l = 0; l < bin_mdef_n_ciphone(mdef); l++) {
-                    p = bin_mdef_phone_id_nearest(mdef, (s3cipid_t) b,
-                                              (s3cipid_t) l, (s3cipid_t) r,
-                                              WORD_POSN_BEGIN);
-                    dict2pid->ldiph_lc[b][r][l] = bin_mdef_pid2ssid(mdef, p);
+        dict2pid->rssid = (xwdssid_t **) ckd_calloc(mdef->n_ciphone, sizeof(xwdssid_t *));
+        dict2pid->lrssid = (xwdssid_t **) ckd_calloc(mdef->n_ciphone, sizeof(xwdssid_t *));
+
+        tmpcimap = ckd_calloc(mdef->n_ciphone, sizeof(s3cipid_t));
+
+        for (b = 0; b < mdef->n_ciphone; ++b) {
+            tmpcimap[b] = 0;
+        }
+
+        /* Always use CI phone SSID, context doesn't matter */
+        for (b = 0; b < mdef->n_ciphone; ++b) {
+            tmpssid = bin_mdef_pid2ssid(mdef, b);
+            dict2pid->ldiph_lc[b][0][0] = tmpssid;
+            dict2pid->lrdiph_rc[b][0][0] = tmpssid;
+
+            dict2pid->rssid[b] = (xwdssid_t *) ckd_calloc(1, sizeof(xwdssid_t));
```

Comment:
Could you explain why do you need to allocate array if in CI mode you just need to return BAD_SSID? I believe you can simply fallback to default value without complex array allocation.

Reply:
There is `dict2pid_rssid(d,ci,lc)` macro that is expected to return `xwdssid_t` structure from `rssid` 2d array. I thought it would be too complex to construct this structure in case of CI mode right in the macro, so I decided to pre-construct and to store it in already existing `rssid` element. Other solution would be to introduce `rssid_single` 1d array with just one `xwdssid_t` per each CI phone. Which approach is preferred?